### PR TITLE
ASP-2210 Added the atomic check-then-set booking insertion

### PR DESCRIPTION
**Pull request by dusktreader**
Added logic to do an atomic check-then-set insertion of the new booking.

---

#### Peer Review
Please follow the upstream omnivector documentation concerning
[peer-review guidelines](https://github.com/omnivector-solutions/Documentation/blob/main/Contributing/pr_review_standards.md#peer-review).

### CODE REVIEW

File: backend/lm_backend/api/routes/bookings.py [hunk 1]
```diff
@@ -26,8 +28,61 @@ async def create_booking(
...
+        literal(booking.job_id).label("job_id"),
+        literal(booking.quantity).label("quantity"),
+    ).cte("vars")
+
+    # Subquery to determine if there are enough licenses to be booked
+    exists_subquery = (
+        select(Feature.id)
+        .select_from(Feature)
+        .join(Booking, and_(Feature.id == Booking.feature_id, Booking.feature_id == booking.feature_id))
+        .join(Inventory, Feature.id == Inventory.feature_id)
```
Comment (matheushent) [inline backend/lm_backend/api/routes/bookings.py:51]: I believe one `join` inside the `select_from` and then `where` makes it easier to read and understand

File: backend/lm_backend/api/routes/bookings.py [hunk 2]
```diff
@@ -26,8 +28,61 @@ async def create_booking(
...
+            exists_subquery
+        ),
+    ).returning(Booking)
+
+    async with db_session.begin():
+        try:
+            result = await db_session.execute(insert_query)
+            db_obj = result.scalars().one_or_none()
+        except Exception as e:
+            raise HTTPException(status_code=400, detail=f"Object could not be read: {e}")
```
Comment (matheushent) [inline backend/lm_backend/api/routes/bookings.py:80]: this seems more like a server error than a client error

**Review by matheushent — CHANGES_REQUESTED** [review]
You made a very nice logic. I just have one point for discussion; beyond that, good work here.

File: backend/lm_backend/api/routes/bookings.py [hunk 3]
```diff
@@ -26,8 +28,61 @@ async def create_booking(
...
+    # CTE used to provide the request values to the insert/select query.
+    vars_cte = select(
+        literal(booking.feature_id).label("feature_id"),
+        literal(booking.job_id).label("job_id"),
+        literal(booking.quantity).label("quantity"),
+    ).cte("vars")
+
+    # Subquery to determine if there are enough licenses to be booked
+    exists_subquery = (
+        select(Feature.id)
```
Comment (julianaklulo) [inline backend/lm_backend/api/routes/bookings.py:48]:
```suggestion
        select(Feature.id, Feature.reserved)
```

File: backend/lm_backend/api/routes/bookings.py [hunk 4]
```diff
@@ -26,8 +28,61 @@ async def create_booking(
...
+        .select_from(Feature)
+        .join(Booking, and_(Feature.id == Booking.feature_id, Booking.feature_id == booking.feature_id))
+        .join(Inventory, Feature.id == Inventory.feature_id)
+        .group_by(Feature.id)
+        .having(
+            func.sum(Booking.quantity)
+            + func.max(Inventory.used)
+            + booking.quantity
+            < func.max(Inventory.total)
```
Comment (julianaklulo) [inline backend/lm_backend/api/routes/bookings.py:57]:
```suggestion
            func.sum(Booking.quantity)
            + func.max(Inventory.used)
            + booking.quantity
            + func.max(Feature.reserved)
            <= func.max(Inventory.total)
```

**Review by julianaklulo — CHANGES_REQUESTED** [review]
The formula to calculate if a booking can be made needs to take into consideration the `reserved`, which is the amount of licenses reserved for usage outside the cluster (in desktop environments, for example).
It was previously known as the limit feature, but that name was misleading.

**Comment by dusktreader** [comment]
This was already merged in another PR.

**Pull request closed without merging**